### DISCUSSION TITLE
Supertab & copilot

### DISCUSSION THREAD

**Discussion by EricDriussi** [LazyVim/LazyVim#4254] [Recipes]
I slightly modified the [supertab recipe](https://www.lazyvim.org/configuration/recipes#supertab) to make tab behave more like other editors as far as copilot integration is concerned.

I'm leaving this here since I've seen comments regarding this around the webs.

```lua
{
  "hrsh7th/nvim-cmp",
  ---@param opts cmp.ConfigSchema
  opts = function(_, opts)
    local has_words_before = function()
      unpack = unpack or table.unpack
      local line, col = unpack(vim.api.nvim_win_get_cursor(0))
      return col ~= 0 and vim.api.nvim_buf_get_lines(0, line - 1, line, true)[1]:sub(col, col):match("%s") == nil
    end

    local cmp = require("cmp")

    opts.mapping = vim.tbl_extend("force", opts.mapping, {
      ["<Tab>"] = cmp.mapping(function(fallback)
        local copilot_ok, copilot_suggestion = pcall(require, "copilot.suggestion")
        if vim.snippet.active({ direction = 1 }) then
          vim.schedule(function()
            vim.snippet.jump(1)
          end)
        elseif copilot_ok and copilot_suggestion.is_visible() then
          copilot_suggestion.accept()
        elseif has_words_before() then
          cmp.complete()
        else
          fallback()
        end
      end, { "i", "s" }),
      ["<S-Tab>"] = cmp.mapping(function(fallback)
        local copilot_ok, copilot_suggestion = pcall(require, "copilot.suggestion")
        if vim.snippet.active({ direction = -1 }) then
          vim.schedule(function()
            vim.snippet.jump(-1)
          end)
        elseif copilot_ok and copilot_suggestion.is_visible() then
          copilot_suggestion.next()
        else
          fallback()
        end
      end, { "i", "s" }),
    })
  end,
}
```

This implies not using tab for selecting the completion suggestion, although I'm guessing it could be adjusted for that purpose as well.

I'm also not 100% sure if the copilot conditional for "Tab" should go before or after the `has_words_before()` logic. Seems to work for me :shrug:

**Comment by chrisgeo** [comment]
I'm curious too - including how to tone down the aggressive position, and recommendations. 

Before using Copilot.lua using the official (neo)vim client it was fairly straight forward.

**Reply by aeciolevy** [reply]
Exactly, before it was pretty straightforward. I am struggling now to get the work pretty robust.

I opened this discussion for it: https://github.com/LazyVim/LazyVim/discussions/4830#discussion-7507424

**Reply by chrisgeo** [reply]
Looks like it worked for you. I didn’t have any on end. Recently they updated a bunch of the frameworks so that may have helped.
